wsmux/session.go

```diff
@@ -0,0 +1,334 @@
+	} else {
+		s.keepAliveInterval = conf.KeepAliveInterval
+	}
+
+	if conf.StreamAcceptDeadline == 0 {
+		s.streamAcceptDeadline = defaultStreamAcceptDeadline
+	} else {
+		s.streamAcceptDeadline = conf.StreamAcceptDeadline
+	}
+
+	if server {
+		s.nextID = 0
+	} else {
+		s.nextID = 1
+	}
+
+	if conf.Log == nil {
+		s.logger = &nilLogger{}
+	} else {
+		s.logger = conf.Log
+	}
+
+	s.conn.SetCloseHandler(s.closeHandler)
+
+	go s.sendLoop()
+	go s.recvLoop()
+	return s
+}
+
+// Accept is used to accept an incoming stream
+func (s *Session) Accept() (net.Conn, error) {
+
+	select {
+	case <-s.closed:
+		return nil, ErrSessionClosed
+	case str := <-s.streamCh:
+		return str, nil
+	}
+}
+
+// Open creates a new stream
+func (s *Session) Open() (net.Conn, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	select {
+	case <-s.closed:
+		return nil, ErrSessionClosed
+	default:
+	}
+
+	id := s.nextID
+	// increment here so that we can wait safely
+	s.nextID += 2
+	if _, ok := s.streams[id]; ok {
+		return nil, ErrDuplicateStream
+	}
+
+	str := newStream(id, s)
+
+	select {
+	case s.writes <- newSynFrame(id):
+		s.streams[id] = str
+	default:
+		// decrement nextID so it can be used by a different stream
+		s.nextID -= 2
+		return nil, ErrBrokenPipe
+	}
+	// unlock mutex and wait
+	s.mu.Unlock()
+
+	select {
+	case <-str.accepted:
+		s.mu.Lock()
+		return str, nil
+	case <-s.closed:
+		s.mu.Lock()
+		// state of s.nextID doesn't matter here
+		delete(s.streams, id)
+		return nil, ErrSessionClosed
+	case <-time.After(s.streamAcceptDeadline):
+		s.mu.Lock()
+		// nextID can be cyclically reused, and previous instance
+		// may be in use by a different stream
+		delete(s.streams, id)
+		return nil, ErrAcceptTimeout
+	}
+
+}
+
+// Close closes the current session and underlying connection
+func (s *Session) Close() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// Check if channel has been closed
+	select {
+	case <-s.closed:
+		return ErrSessionClosed
+	default:
+	}
+
+	close(s.closed)
+	if s.closeConn {
+		_ = s.conn.Close()
+	}
+
+	for _, v := range s.streams {
+		_ = v.Close()
+	}
+
+	return nil
+}
+
+// Addr used for implementing net.Listener
+func (s *Session) Addr() net.Addr {
+	return s.conn.LocalAddr()
+}
+
+func (s *Session) closeHandler(code int, text string) error {
+	s.logger.Printf("ws conn closed: code %d : %s", code, text)
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.closeConn = false
+	return s.Close()
+}
+
+func (s *Session) sendLoop() {
+	for {
+		select {
+		case <-s.closed:
+			return
+
+		case fr, ok := <-s.writes:
+			if !ok {
+				s.logger.Print("write channel closed")
+				return
+			}
+			err := s.conn.WriteMessage(websocket.BinaryMessage, fr.Write())
+			s.logger.Printf("wrote message: %v", fr)
+			if err != nil {
+				s.logger.Print(err)
+			}
+		}
+	}
+}
+
+func (s *Session) recvLoop() {
+	for {
+		select {
+		case <-s.closed:
+			return
+		default:
+		}
+
+		t, msgReader, err := s.conn.NextReader()
+		if err != nil {
+			_ = s.Close()
```

Comment:
Instead of doing this... I suggest that you make a internal method:
`s.abort(err error)`

Then whenever, a fatal error happens you call `s.abort(err)`...
From there it is just a matter of making sure that `abort` always releases all resources, closes all streams, ensures nothing deadlocks and that functions correctly return errors...

I would do this by creating:
```go
type Session struct {
  m sync.Mutex
  closed chan struct{}
  acceptErr error
}

// abort the session with given error
func (s *Session) abort(err error) error {
  s.m.Lock()
  defer s.m.Unlock()

  if s.isClosed() {
    return
  }

  // Setup state to mark this as resolved
  s.acceptErr = err

  // Only mark as closed at the end
  close(s.closed)
}

// isClosed returns true, if session is closed
func (s *Session) isClosed() bool {
  select {
    case <- s.closed:
      return true
    default:
      return false
  }
}

func (s *Session) Accept() (net.Conn, error) {

	select {
	case <-s.closed:
		return nil, s.acceptErr
	case str := <-s.streamCh:
		return str, nil
	}
}

```